src/main/java/uk/gov/companieshouse/chs/gov/uk/notify/integration/api/templatepersonalisation/TemplatePersonaliser.java

```diff
@@ -0,0 +1,104 @@
+                                            Map<String, String> personalisationDetails,
+                                            Address address) {
+
+        var context = new Context();
+
+        // Use today's date for traceability.
+        var format = DateTimeFormatter.ofPattern("dd MMMM yyyy");
+        context.setVariable("date", LocalDate.now().format(format));
+
+        var upperCaseCompanyName = getUpperCasedCompanyName(personalisationDetails);
+        populateAddress(context, address, upperCaseCompanyName);
+        personaliseLetter(context, personalisationDetails, upperCaseCompanyName);
+
+        validator.validateContextForTemplate(context, template);
+
+        return templateEngine.process(template.id(), context);
+    }
+
+    private String getUpperCasedCompanyName(Map<String, String> personalisationDetails) {
+        // Company name must be provided and is always rendered in UPPER CASE in the letter.
+        if (isEmpty(personalisationDetails.get("company_name"))) {
+            throw new LetterValidationException(
+                    "No company name found in the letter personalisation details.");
+        }
+        return personalisationDetails.get("company_name").toUpperCase();
+    }
+
+    @SuppressWarnings("java:S1135") // TODO left in place intentionally for now.
+    private void populateAddress(Context context, Address address, String upperCaseCompanyName) {
+        context.setVariable("address_line_1",
+                uppercaseIfCompanyName(address.getAddressLine1(), upperCaseCompanyName));
+        context.setVariable("address_line_2",
+                uppercaseIfCompanyName(address.getAddressLine2(), upperCaseCompanyName));
+        context.setVariable("address_line_3",
+                uppercaseIfCompanyName(address.getAddressLine3(), upperCaseCompanyName));
+        context.setVariable("address_line_4",
+                uppercaseIfCompanyName(address.getAddressLine4(), upperCaseCompanyName));
+        context.setVariable("address_line_5",
+                uppercaseIfCompanyName(address.getAddressLine5(), upperCaseCompanyName));
+        context.setVariable("address_line_6",
+                uppercaseIfCompanyName(address.getAddressLine6(), upperCaseCompanyName));
+        // TODO DEEP-287 postcode_or_country or just line 7?
+        // Consider populating this field with the last populated address line...
+        context.setVariable("postcode_or_country", address.getAddressLine7());
+    }
+
+    private void personaliseLetter(Context context,
+                                   Map<String, String> personalisationDetails,
+                                   String upperCaseCompanyName) {
+        personalisationDetails.keySet().forEach(name ->
+                context.setVariable(name,
+                        uppercaseIfCompanyName(personalisationDetails.get(name),
+                                upperCaseCompanyName)));
+    }
```

Comment:
Same idea with returning `Context` here so it's more explicit that it's being changed, also we could maybe use the more intuitive iteration of `forEach`:

```suggestion
    private Context personaliseLetter(Context context,
                                      Map<String, String> personalisationDetails,
                                      String upperCaseCompanyName) {
        personalisationDetails.forEach((key, value) ->
                context.setVariable(key, uppercaseIfCompanyName(value, upperCaseCompanyName)));
        return context;
    }
```